### PR TITLE
Change proxy metrics to conform metrics guidelines

### PR DESCRIPTION
**Pull request by danielqsj**
**What type of PR is this?**

/kind feature
/sig instrumentation

**What this PR does / why we need it**:

As part of [kubernetes metrics overhaul](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0031-kubernetes-metrics-overhaul.md), change proxy metrics to conform [Kubernetes metrics instrumentation guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md).

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

This patch does not remove the existing metrics but mark them as deprecated.
We need 2 releases for users to convert monitoring configuration.

**Does this PR introduce a user-facing change?**:

```release-note
Change proxy metrics to conform metrics guidelines.
The metrics `sync_proxy_rules_latency_microseconds` is deprecated, and will be removed in a future release, please convert to metrics`sync_proxy_rules_latency_seconds`.
```

### CODE REVIEW

File: pkg/proxy/winkernel/metrics.go [hunk 1]
```diff
@@ -27,10 +27,19 @@ const kubeProxySubsystem = "kubeproxy"
...
+			Buckets:   prometheus.ExponentialBuckets(0.001, 2, 15),
+		},
+	)
+
+	DeprecatedSyncProxyRulesLatency = prometheus.NewHistogram(
 		prometheus.HistogramOpts{
 			Subsystem: kubeProxySubsystem,
 			Name:      "sync_proxy_rules_latency_microseconds",
-			Help:      "SyncProxyRules latency",
+			Help:      "SyncProxyRules latency in microseconds",
```
Comment (logicalhan) [inline pkg/proxy/winkernel/metrics.go]: Should we output the deprecation notice in the help?

**Review by logicalhan — APPROVED** [review]
/lgtm

File: pkg/proxy/winkernel/metrics.go (hunk 1 above)
Comment (danielqsj) [reply pkg/proxy/winkernel/metrics.go]: Great idea. This kind of notice should be applied in all deprecation metrics. I will mark all of them later.

**Comment by danielqsj** [comment]
@brancz @logicalhan
Added `Deprecated` in metrics description. PTAL

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -643,7 +643,8 @@ func (proxier *Proxier) syncProxyRules() {
 
 	start := time.Now()
 	defer func() {
-		metrics.SyncProxyRulesLatency.Observe(metrics.SinceInMicroseconds(start))
+		metrics.SyncProxyRulesLatency.Observe(metrics.SinceInSeconds(start))
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Is this dropping precision? why?  I expect that most of the time this takes < 1 second, butthe difference between 0.2 seconds and 0.9 seconds is kind of meaningful.  It seems to be float so maybe it's just weird naming?

**Review by thockin — COMMENTED** [review]
One question to make sure I understand.  assuming that is right, LGTM

File: pkg/proxy/iptables/proxier.go (hunk 2 above)
Comment (danielqsj) [reply pkg/proxy/iptables/proxier.go]: @thockin In fact, `SinceInSeconds` (nanoseconds level) have higher precision than `SinceInMicroseconds` (microseconds level).
For example:
The result of this case:
```
fmt.Printf("SinceInMicroseconds: %v\nSinceInSeconds: %v\n", SinceInMicroseconds(time.Now().Add(-time.Second)), SinceInSeconds(time.Now().Add(-time.Second)))
```
is
```
SinceInMicroseconds: 1.000001e+06
SinceInSeconds: 1.000000806
```
The name `SinceInSeconds` is used to show its unit as `seconds`.
But we can discuss about this naming. cc @brancz

Comment (brancz) [reply pkg/proxy/iptables/proxier.go]: We're not loosing precision as it's still a float64, this is just a consistency change, to make sure all metrics use the base unit, which is the best practice so it is easier for different metrics to be used with each other (this is part of the Prometheus instrumentation best practices, which among others includes using [seconds as time base unit](https://prometheus.io/docs/practices/naming/#base-units). Complying to the Prometheus best practices is one of the motivations for the [metrics overhaul KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0031-kubernetes-metrics-overhaul.md#motivation), which this is part of.

**Comment by brancz** [comment]
@thockin as the KEP mentions:

> While uncertain at this point, once cleaned up, this effort may put us a step closer to having stability guarantees for Kubernetes around metrics. Currently metrics are excluded from any kind of stability requirements.

We absolutely want to have deprecation/stability guarantees around metrics, but are not comfortable doing so at the moment. This is sig-instrumentation's plan however, and after we've done the cleanup and put tooling in place to ensure all metrics do comply to the guidelines we intend to put stability guarantees in place.

**Comment by brancz** [comment]
@thockin ping :slightly_smiling_face:

**Comment by thockin** [comment]
I look forward to a discussion of deprecation of metrics :)

/lgtm
/approve

**Pull request merged**